common/src/Model/Entity.cpp

```diff
@@ -150,30 +150,31 @@ namespace TrenchBroom {
         }
 
         void Entity::doChildWasAdded(Node* node) {
-            nodeBoundsDidChange();
+            nodeBoundsDidChange(bounds());
         }
         
         void Entity::doChildWasRemoved(Node* node) {
-            nodeBoundsDidChange();
+            nodeBoundsDidChange(bounds());
         }
 
-        void Entity::doNodeBoundsDidChange() {
+        void Entity::doNodeBoundsDidChange(const BBox3& oldBounds) {
             invalidateBounds();
         }
         
-        void Entity::doChildBoundsDidChange(Node* node) {
-            nodeBoundsDidChange();
+        void Entity::doChildBoundsDidChange(Node* node, const BBox3& oldBounds) {
+            const BBox3& myBounds = bounds();
+            if (!myBounds.encloses(oldBounds) && !myBounds.encloses(node->bounds())) {
+                // Our bounds will change only if the child's bounds potentially contributed to our own bounds.
+                nodeBoundsDidChange(myBounds);
+            }
         }
 
         bool Entity::doSelectable() const {
             return !hasChildren();
         }
 
         void Entity::doPick(const Ray3& ray, PickResult& pickResult) const {
-            if (hasChildren()) {
-                for (const Node* child : Node::children())
-                    child->pick(ray, pickResult);
-            } else {
+            if (!hasChildren()) {
```

Comment:
Just a question, why don't you need to call pick() on the children anymore?

Reply:
Because the children are now all added to the node tree, which they were not before.